### PR TITLE
added JWS support for switch generated msg

### PR DESCRIPTION
**Pull request by shashi165**
Added the support for JWS signing for switch generated messages

### CODE REVIEW

File: secrets/jwsSigningKey.key [hunk 1]
```diff
@@ -0,0 +1,3 @@
+-----BEGIN PRIVATE KEY-----
+some valid signature
```
Comment (mdebarros) [inline secrets/jwsSigningKey.key:2]: Can we not put a dev/test key here?

I.e. what happens if JWS signage is enabled and this default non-valid key is used?

File: src/model/quotes.js [hunk 2]
```diff
@@ -1102,6 +1126,31 @@ class QuotesModel {
...
+  generateRequestHeadersForJWS (headers, noAccept) {
+    const ret = {
+      'Content-Type': 'application/vnd.interoperability.quotes+json;version=1.0',
+      date: headers.date,
+      'fspiop-source': headers['fspiop-source'],
+      'fspiop-destination': headers['fspiop-destination'],
+      'fspiop-http-method': headers['fspiop-http-method'],
+      'fspiop-signature': headers['fspiop-signature'],
+      'fspiop-uri': headers['fspiop-uri'],
+      'User-Agent': null, // yuck! node-fetch INSISTS on sending a user-agent header!? infuriating!
```
Comment (mdebarros) [inline src/model/quotes.js:1143]: I do not think this is applicable since I see that axios is being used?

File: secrets/jwsSigningKey.key (hunk 1 above)
Comment (bushjames) [reply secrets/jwsSigningKey.key:2]: I agree. I think we should use a self generated default key pair and also, put the public key in the sdk-scheme-adapter image. Update docs to make it imperative for users/implementers to change the defaults in their own installations. We could include an openssl command to generate a new keypair.

File: src/model/quotes.js (hunk 2 above)
Comment (shashi165) [reply src/model/quotes.js]: removed

**Comment by millerabel** [comment]
Hi James, good ideas to make this simple to use in eval environments. However, would it be preferable If there was no default and that failure were induced unless the key generation we’re done by implementers? Seems safer.

But that means including a test key for, er, testing.

Could the system assert that the test key is not used outside of the regression environment? E.g., the test key is trivially generated from a human- and machine-recognizable sequence, is repeatable, and is easily detected in code. This would ensure repeatable encryption of test traffic, which is both good for testing somethings and bad for others. 

Or that a new test key is generated each time test environment is built? This is the most robust but is unrepeatable.

— Miller

On May 11, 2020, at 2:41 AM, James Bush <notifications@github.com> wrote:

﻿
@bushjames commented on this pull request.

In secrets/jwsSigningKey.key:

> @@ -0,0 +1,3 @@
+-----BEGIN PRIVATE KEY-----
+some valid signature
I agree. I think we should use a self generated default key pair and also, put the public key in the sdk-scheme-adapter image. Update docs to make it imperative for users/implementers to change the defaults in their own installations. We could include an openssl command to generate a new keypair.

**Comment by bushjames** [comment]
Hi @millerabel ! I suppose the code could check an env var e.g. NODE_ENV to look for "production" and throw at startup if the test keys are installed. We would have to have a mechanism for manipulating that var in a safe way....the helm chart defaults could set it to "development", but I cant see a way to force people to change it in production.

**Comment by msk-** [comment]
Perhaps the service ought to fail instructively when the configuration is inconsistent. A pseudo-implementation:
```
if (jws_enabled) {
  try {
    load_jws();
  } catch () {
    log('You have enabled JWS signing but have not set up your key correctly. Follow the steps at https://github.com/mojaloop/quoting-service#configure-jws to configure the service correctly.');
    process.exit(1);
  }
}
```

**Comment by bushjames** [comment]
@msk- indeed. The bigger problem is in how we make it easy for people to get JWS out-of-the-box with zero config required for first time users (we are thinking a pre-loaded dev key), and at the same time make sure nobody uses the dev key in production.

**Comment by msk-** [comment]
I understand. Usability and security are pretty commonly in tension though. If we point a user directly at the documentation, and that contains a series of easy steps for the user to take, that seems a reasonable compromise.

**Comment by millerabel** [comment]
@James, Perhaps it’s better to require it to be generated as part of test config and deployment? So when helm charts are “parameterized” during deployment, With local values, the key must be generated to avoid failure of acceptance tests?

If a pattern for ensuring key pairs are generated as part of deployment can be developed, and applied to other areas of the system, it would avoid the same kind of configuration weakness from creeping in elsewhere and that might easily go unnoticed during deployment.

— Miller

On May 12, 2020, at 6:41 AM, James Bush <notifications@github.com> wrote:

﻿
Hi @millerabel ! I suppose the code could check an env var e.g. NODE_ENV to look for "production" and throw at startup if the test keys are installed. We would have to have a mechanism for manipulating that var in a safe way....the helm chart defaults could set it to "development", but I cant see a way to force people to change it in production.

**Comment by bushjames** [comment]
OK, if you are comfortable with pods not starting if JWS is enabled and a key is not installed then I am also OK with that. We _may_ be able to use helm to require the key being specified in values if JWS is enabled. Then the deployment would fail if the config was mismatched.

> @james, Perhaps it’s better to require it to be generated as part of test config and deployment? So when helm charts are “parameterized” during deployment, With local values, the key must be generated to avoid failure of acceptance tests? If a pattern for ensuring key pairs are generated as part of deployment can be developed, and applied to other areas of the system, it would avoid the same kind of configuration weakness from creeping in elsewhere and that might easily go unnoticed during deployment. — Miller On May 12, 2020, at 6:41 AM, James Bush <notifications@github.com> wrote: ﻿ Hi @millerabel ! I suppose the code could check an env var e.g. NODE_ENV to look for "production" and throw at startup if the test keys are installed. We would have to have a mechanism for manipulating that var in a safe way....the helm chart defaults could set it to "development", but I cant see a way to force people to change it in production. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by millerabel** [comment]
I think it’s worth iterating a bit to find the right balance of safety and convenience. In principle, catching this in Helm values (along with other required configuration) seems like a good idea. However, as keys move into a KMS, it isn’t clear how that would work. I suspect the POD would just fail if it could not retrieve a JWS session key from the KMS, and that configuration of the KMS is a separate matter. MKS suggestion to have the JWS setup throw an instructional suggestion also seems like a good idea.


On May 12, 2020, at 9:43 AM, James Bush <notifications@github.com> wrote:


OK, if you are comfortable with pods not starting if JWS is enabled and a key is not installed then I am also OK with that. We may be able to use helm to require the key being specified in values if JWS is enabled. Then the deployment would fail if the config was mismatched.

@james <https://github.com/james>, Perhaps it’s better to require it to be generated as part of test config and deployment? So when helm charts are “parameterized” during deployment, With local values, the key must be generated to avoid failure of acceptance tests? If a pattern for ensuring key pairs are generated as part of deployment can be developed, and applied to other areas of the system, it would avoid the same kind of configuration weakness from creeping in elsewhere and that might easily go unnoticed during deployment. — Miller On May 12, 2020, at 6:41 AM, James Bush notifications@github.com <mailto:notifications@github.com> wrote: ﻿ Hi @millerabel <https://github.com/millerabel> ! I suppose the code could check an env var e.g. NODE_ENV to look for "production" and throw at startup if the test keys are installed. We would have to have a mechanism for manipulating that var in a safe way....the helm chart defaults could set it to "development", but I cant see a way to force people to change it in production. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by msk-** [comment]
@bushjames 

We certainly can prevent a deployment due to inconsistent configuration. Example: https://github.com/mojaloop/helm/blob/caa91a81922ecc7b8344286ff882e97300ea6409/mojaloop-simulator/templates/config-jws-public-keys.yaml#L16-L18

We _might_ be able to generate JWT keys. Here's an example generating some keys for mTLS: https://github.com/mojaloop/helm/blob/caa91a81922ecc7b8344286ff882e97300ea6409/mojaloop-simulator/templates/secret.yaml#L35-L40

Whether we can generate keys I think will depend on the hash function used by Sprig crypto (a lib Helm exposes to template users). Could read the source, or try it and see. Or both.

**Comment by bushjames** [comment]
@shashi165 please can you have a go at this. Revert having a default key and "require" one in helm as per Matts examples above so if JWS is enabled the deploy will fail if the user does not generate and install one in the values file. Many thanks,

**Comment by shashi165** [comment]
@bushjames - @msk- has raise a PR for the helm change, that will make it mandatory to generate the key, if someone enables the JWS signing. It will ignore the default key. I will keep the default key for unit testing.

**Pull request merged**